File: 2434.using-a-robot-to-print-the-lexicographically-smallest-string.py
#
# @lc app=leetcode id=2434 lang=python3
#
# [2434] Using a Robot to Print the Lexicographically Smallest String
#

# @lc code=start
from collections import Counter
from string import ascii_lowercase
class Solution:
    def robotWithString(self, s: str) -> str:


        #! how to ensure the lexicographically smallest string?
        #! when should the letter of t be added to paper?

        #! how to keep record of the smallest character of the rest of the string s?
        stack, ans = [], []
        #Record the count of each character
        cnt = Counter(s)
        #Record the smallest character of the rest of the string s
        min = 0

        for c in s:
            cnt[c] -= 1
            while min < 25 and cnt[ascii_lowercase[min]] == 0:
                min += 1  
            stack.append(c)
            while stack and stack[-1] <= ascii_lowercase[min]:
                ans.append(stack.pop())
        return ''.join(ans) 
        
# @lc code=end

File: 2434.using-a-robot-to-print-the-lexicographically-smallest-string.java
/*
 * @lc app=leetcode id=2434 lang=java
 *
 * [2434] Using a Robot to Print the Lexicographically Smallest String
 */

// @lc code=start
import java.util.Stack;
class Solution {
    public String robotWithString(String s) {
        //use prefix to store the smallest character from the right

        //Again the idea of using prefix
        int n = s.length();
        char[] prefix = new char[n + 1];
        prefix[n] = Character.MAX_VALUE;
        for(int i = n - 1; i >= 0; i--){
            prefix[i] = (char) Math.min(s.charAt(i), prefix[i + 1]);
        }
        StringBuilder ans = new StringBuilder();
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < n; i++){
            stack.push(s.charAt(i));
            while(!stack.isEmpty() && stack.peek() <= prefix[i + 1]){
                ans.append(stack.pop());
            }
        }
        return ans.toString();

     }
}
// @lc code=end

